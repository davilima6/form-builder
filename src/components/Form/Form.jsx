// @flow
import * as React from 'react';
import styled, { css } from 'styled-components';
import { Error as TypeError, hasErrors, normalize } from '../../utils';
import { Button, Field } from '..';

const { useCallback } = React;

const StyledForm = styled.form`
  background-color: #fff;
  border: 1px solid var(--lite-grey);
  border-radius: 4px;
  box-shadow: 0 2px 2px 0 rgba(47, 55, 64, 0.2), 0 6px 10px 0 rgba(47, 55, 64, 0.2);
  margin: 1em 0 1.5em;
  min-width: 60vw;
  padding: 2em 2em 1em;

  ${props => props.error && 'border: 2px solid var(--dark-red'};
`;

const FormHeader = styled.div`
  margin: 0 0 2em 1em;
`;

const FormTitle = styled.h2``;

const FormDescription = styled.p`
  font-style: italic;
`;

const FormActions = styled.div`
  display: flex;
  margin: 3em 0.5em 0;
  padding: 0.5em;
`;

const StyledError = styled.div`
  background-color: var(--lite-red);
  border-radius: 3px;
  box-sizing: border-box;
  flex: 0 0 100%;
  margin-top: 1em;
  padding: 1em;
`;

type Props = {
  schema: Array<Field>,
  errors?: { [string]: Array<TypeError> },
  errorMessages?: { [string]: Array<string> },
  data?: { [string]: string },
  dispatch: Function,
  onSubmit?: Function,
  onCancel?: Function,
};

const Form = ({
  schema,
  errors = {},
  errorMessages,
  data,
  dispatch,
  onSubmit,
  onCancel,
}: Props) => {
  const onChangeField = useCallback((label, value) => {
    dispatch({ type: 'update', payload: { [normalize(label)]: value } });
  }, []);

  const FormHead = () => (
    <FormHeader>
      <FormTitle>Title of the form</FormTitle>
      <FormDescription>Longer description of the form</FormDescription>
    </FormHeader>
  );

  return (
    <StyledForm method="post" onSubmit={onSubmit} error={errors.length}>
      <FormHead />

      {hasErrors(errors) && <StyledError>{errorMessages.hasErrors()}</StyledError>}

      {schema.map((field) => {
        const key = normalize(`${field.type}_${field.label}`);
        const fieldName = normalize(field.label);

        return (
          <Field
            key={key}
            id={fieldName}
            label={field.label}
            type={field.type}
            value={data[fieldName] ? data[fieldName] : ''}
            errors={errors[fieldName]}
            errorMessages={errorMessages[fieldName]}
            onChangeField={onChangeField}
          />
        );
      })}
      <FormActions>
        {onSubmit && <Button type="submit" title="Send" floated="right" primary />}
        {onCancel && <Button type="reset" title="Clear" floated="left" onClick={onCancel} />}
      </FormActions>
    </StyledForm>
  );
};

Form.defaultProps = {
  data: {},
  errors: {},
  errorMessages: {},
  onSubmit: () => {},
  onCancel: () => {},
};

export default Form;
export { StyledError };
