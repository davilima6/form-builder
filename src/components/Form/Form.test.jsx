import React from 'react';
import { cleanup, render } from 'react-testing-library';
import 'jest-dom/extend-expect';
import 'jest-styled-components';
import Form from './Form';

describe('Form', () => {
  let component;

  beforeEach(() => {
    cleanup();
    component = render(<Form />);
  });

  test('renders without crashing', () => expect(component).toBeDefined());

  test('matches snapshot', () => {
    const { asFragment } = component;

    expect(asFragment()).toMatchSnapshot();
  });
});
