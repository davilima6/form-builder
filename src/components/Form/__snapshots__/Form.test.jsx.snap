// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Form matches snapshot 1`] = `
<DocumentFragment>
  <form
    class="sc-fjdhpX laQhmB"
    method="post"
  >
    <div
      class="sc-jzJRlG HjlPv"
    >
      <h2
        class="sc-cSHVUG CAuLq"
      >
        Title of the form
      </h2>
      <p
        class="sc-kAzzGY idFEFu"
      >
        Longer description of the form
      </p>
    </div>
    <div
      class="sc-bZQynM EsGCe"
    >
      <div
        class="sc-dnqmqq jvzWdJ"
      >
        <label
          class="sc-EHOje jvzENE"
          for="field-first-name"
        >
          <div
            class="sc-htoDjs dhLPgd"
          >
            First Name
          </div>
          <input
            class="sc-gzVnrw eWqNKI"
            id="field-first-name"
            name="first-name"
            type="text"
            value=""
          />
        </label>
      </div>
    </div>
    <div
      class="sc-bZQynM EsGCe"
    >
      <div
        class="sc-dnqmqq jvzWdJ"
      >
        <label
          class="sc-EHOje jvzENE"
          for="field-last-name"
        >
          <div
            class="sc-htoDjs dhLPgd"
          >
            Last Name
          </div>
          <input
            class="sc-gzVnrw eWqNKI"
            id="field-last-name"
            name="last-name"
            type="text"
            value=""
          />
        </label>
      </div>
    </div>
    <div
      class="sc-bZQynM EsGCe"
    >
      <div
        class="sc-dnqmqq jvzWdJ"
      >
        <label
          class="sc-EHOje jvzENE"
          for="field-email"
        >
          <div
            class="sc-htoDjs dhLPgd"
          >
            email
          </div>
          <input
            class="sc-gzVnrw eWqNKI"
            id="field-email"
            name="email"
            placeholder="your username@google.com"
            type="email"
            value=""
          />
        </label>
      </div>
    </div>
    <div
      class="sc-bZQynM EsGCe"
    >
      <div
        class="sc-dnqmqq jvzWdJ"
      >
        <label
          class="sc-EHOje jvzENE"
          for="field-age"
        >
          <div
            class="sc-htoDjs dhLPgd"
          >
            Age
          </div>
          <input
            class="sc-gzVnrw eWqNKI"
            id="field-age"
            name="age"
            type="number"
            value=""
          />
        </label>
      </div>
    </div>
    <div
      class="sc-bZQynM EsGCe"
    >
      <div
        class="sc-ifAKCX cZFrkc sc-dnqmqq jvzWdJ"
      >
        <div
          class="sc-htpNat iknpIM sc-htoDjs dhLPgd"
        >
          Gender
        </div>
        <label
          class="sc-bwzfXH dnzClS sc-EHOje jvzENE"
          for="field-gender-female"
        >
          <div
            class="sc-bxivhb kTYxlA"
          >
            Female
          </div>
          <input
            class="sc-gzVnrw eWqNKI"
            id="field-gender-female"
            name="gender"
            type="radio"
            value="0"
          />
        </label>
        <label
          class="sc-bwzfXH dnzClS sc-EHOje jvzENE"
          for="field-gender-male"
        >
          <div
            class="sc-bxivhb kTYxlA"
          >
            Male
          </div>
          <input
            class="sc-gzVnrw eWqNKI"
            id="field-gender-male"
            name="gender"
            type="radio"
            value="1"
          />
        </label>
        <label
          class="sc-bwzfXH dnzClS sc-EHOje jvzENE"
          for="field-gender-non-binary"
        >
          <div
            class="sc-bxivhb kTYxlA"
          >
            Non-Binary
          </div>
          <input
            class="sc-gzVnrw eWqNKI"
            id="field-gender-non-binary"
            name="gender"
            type="radio"
            value="2"
          />
        </label>
      </div>
    </div>
    <div
      class="sc-bZQynM EsGCe"
    >
      <div
        class="sc-dnqmqq jvzWdJ"
      >
        <label
          class="sc-EHOje jvzENE"
          for="field-date-of-birth"
        >
          <div
            class="sc-htoDjs dhLPgd"
          >
            Date of Birth
          </div>
          <input
            class="sc-gzVnrw eWqNKI"
            id="field-date-of-birth"
            name="date-of-birth"
            type="date"
            value=""
          />
        </label>
      </div>
    </div>
    <div
      class="sc-chPdSV cxboSq"
    >
      <button
        class="sc-bdVaJa bKWFes"
        disabled=""
        title="Form has not been filled out yet so it cannot be submitted"
        type="submit"
      >
        Send
      </button>
      <button
        class="sc-bdVaJa fYbgcP"
        title="Clear"
        type="reset"
      >
        Clear
      </button>
    </div>
  </form>
</DocumentFragment>
`;
